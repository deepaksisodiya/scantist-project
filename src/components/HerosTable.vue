<template>
  <div>
    <input v-model="searchName" />
    <div v-if="heros.isLoading">
      Loading...
    </div>
    <div v-else-if="heros.isError">
      Error in fetching heros
    </div>
    <table v-else>
      <thead>
        <tr>
          <th>Hero Image</th>
          <th>Hero Name</th>
          <th>PRO Win Rate</th>
          <th>Pro Pick Count</th>
          <th>Pro Ban Count</th>
        </tr>
      </thead>
      <tbody>
        <tr v-for="hero in heros.filterData" v-bind:key="hero.id">
          <td><img :src="`http://cdn.dota2.com/${hero.img}`" /></td>
          <td>{{ hero.localized_name }}</td>
          <td>{{ hero.pro_win }}</td>
          <td>{{ hero.pro_pick }}</td>
          <td>{{ hero.pro_ban }}</td>
        </tr>
      </tbody>
    </table>
  </div>
</template>

<script>
import axios from "axios";
import debounce from "lodash.debounce";

export default {
  name: "HerosTable",

  data: function() {
    return {
      heros: {
        isLoading: false,
        isError: false,
        data: [],
        filterData: []
      },

      searchName: null
    };
  },

  watch: {
    searchName: function() {
      this.search(this.searchName, this);
    }
  },

  methods: {
    search: debounce((search, vm) => {
      const self = vm;
      self.filterHeros(self);
    }, 500),

    filterHeros() {
      const filterData = this.heros.data.filter(hero => {
        const localizedNameLowerCase = hero.localized_name.toLowerCase();
        const searchNameLowerCase = this.searchName.toLowerCase();
        return localizedNameLowerCase.includes(searchNameLowerCase);
      });
      this.heros.filterData = filterData;
    }
  },

  async mounted() {
    this.heros.isLoading = true;
    this.heros.isError = false;

    try {
      const response = await axios.get(
        "https://api.opendota.com/api/heroStats"
      );
      if (response.status === 200) {
        this.heros.data = response.data;
        this.heros.filterData = response.data;
      }
    } catch (error) {
      this.heros.isError = true;
    } finally {
      this.heros.isLoading = false;
    }
  }
};
</script>
