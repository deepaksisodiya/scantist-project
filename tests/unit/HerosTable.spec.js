import { shallowMount } from "@vue/test-utils";

import HerosTable from "./../../src/components/HerosTable.vue";

describe("HerosTable", () => {
  let store;

  beforeEach(() => {
    store = {
      state: {
        isLoading: false,
        isError: false,
        heros: []
      },
      getters: {
        filterHerosByName: () => {}
      },
      actions: {
        fetchHeros: () => {}
      },
      dispatch: () => {}
    };
  });

  test("Test for isLoading condition", () => {
    store.state.isLoading = true;

    const wrapper = shallowMount(HerosTable, {
      mocks: {
        $store: store
      }
    });

    expect(wrapper).toMatchSnapshot();
  });

  test("Test for isError condition", () => {
    store.state.isError = true;

    const wrapper = shallowMount(HerosTable, {
      mocks: {
        $store: store
      }
    });

    expect(wrapper).toMatchSnapshot();
  });

  test("Test for where we have heros data", () => {
    store.state.heros = [
      {
        id: 1,
        img: "/apps/dota2/images/heroes/antimage_full.png?",
        localized_name: "Deepak",
        pro_win: 34,
        pro_pick: 50,
        pro_ban: 80
      },
      {
        id: 2,
        img: "apps/dota2/images/heroes/axe_full.png?",
        localized_name: "Piyush",
        pro_win: 44,
        pro_pick: 52,
        pro_ban: 60
      },
      {
        id: 3,
        img: "/apps/dota2/images/heroes/bane_full.png?",
        localized_name: "Ankur",
        pro_win: 32,
        pro_pick: 23,
        pro_ban: 45
      },
      {
        id: 4,
        img: "/apps/dota2/images/heroes/bloodseeker_full.png?",
        localized_name: "Mayur",
        pro_win: 40,
        pro_pick: 60,
        pro_ban: 90
      }
    ];

    const wrapper = shallowMount(HerosTable, {
      mocks: {
        $store: store
      }
    });

    expect(wrapper).toMatchSnapshot();
  });
});
