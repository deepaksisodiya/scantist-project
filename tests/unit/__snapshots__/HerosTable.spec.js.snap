// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HerosTable Test for isError condition 1`] = `
<div>
  <el-input placeholder="Search By Name"></el-input>
  <div>
    Error in fetching heros
  </div>
</div>
`;

exports[`HerosTable Test for isLoading condition 1`] = `
<div>
  <el-input placeholder="Search By Name"></el-input>
  <div>
    Loading...
  </div>
</div>
`;

exports[`HerosTable Test for where we have heros data 1`] = `
<div>
  <el-input placeholder="Search By Name"></el-input>
  <el-table>
    <el-table-column label="Hero Image"></el-table-column>
    <el-table-column prop="localized_name" label="Hero Name" width="180"></el-table-column>
    <el-table-column prop="pro_win" label="PRO Win Rate" width="180"></el-table-column>
    <el-table-column prop="pro_pick" label="Pro Pick Count"></el-table-column>
    <el-table-column prop="pro_ban" label="Pro Ban Count"></el-table-column>
  </el-table>
</div>
`;
